core/src/kmx/kmx_plus.cpp

```diff
@@ -46,33 +50,102 @@ validate_section_name(KMX_DWORD ident) {
+/**
+ * @brief cast to a COMP_KMXPLUS_HEADER from bytes
+ *
+ * @param data
+ * @param length
+ * @param ident
+ * @return const kmx::COMP_KMXPLUS_HEADER*
+ */
+static const kmx::COMP_KMXPLUS_HEADER *
+header_from_bytes(const uint8_t *data, KMX_DWORD length, uint32_t ident) {
+  if (!data) {
+    debug_println("!data");
+    return nullptr;
+  }
+  if (length < LDML_LENGTH_HEADER) {
+    debug_println("length < LDML_LENGTH_HEADER");
+  }
+  const COMP_KMXPLUS_HEADER *all = reinterpret_cast<const COMP_KMXPLUS_HEADER *>(data);
+  if (!all->valid(length)) {
+    debug_println("header failed validation");
+    return nullptr;
+  }
+  if (all->ident != ident) {
+    debug_println("header had wrong section id");
+    return nullptr;
+  }
+  return all;
+}
+
+/**
+ * @brief Accessor for a section based on bytes
+ *
+ * @tparam T
+ * @param data
+ * @param length
+ * @return const T*
+ */
+template <class T>
+const T *section_from_bytes(const uint8_t *data, KMX_DWORD length) {
+  if (length < sizeof(T)) { // Does not include dynamic data. First check.
+    debug_println("length < sizeof(section)");
+    return nullptr;
+  }
+  const COMP_KMXPLUS_HEADER *header = header_from_bytes(data, length, T::IDENT);
+  const T *section = reinterpret_cast<const T *>(header);
+  if (section != nullptr && section->valid(length)) {
+    return section;
+  } else {
+    return nullptr;
+  }
+}
+
+template <class T>
+const T *section_from_sect(const COMP_KMXPLUS_SECT* sect) {
+  const uint8_t *rawbytes = reinterpret_cast<const uint8_t *>(sect);
+  if (rawbytes == nullptr) {
+    debug_println("section_from_sect(nullptr) == nullptr");
+    return nullptr;
+  }
+  KMX_DWORD offset = sect->find(T::IDENT);
+  if (!offset) {
+    debug_println("section_from_sect() - not found");
+    return nullptr;
+  }
+  KMX_DWORD entrylength = sect->total - offset;
+  return section_from_bytes<T>(rawbytes+offset, entrylength);
+}
+
+bool
+COMP_KMXPLUS_HEADER::valid(KMX_DWORD length) const {
+  KMXPLUS_PRINTF((": (%X) size 0x%X\n", ident, size));
+  if (size < LDML_LENGTH_HEADER) {
+    debug_println("size < LDML_LENGTH_HEADER");
     return false;
   }
-  if(!validate_section_name(hdr->ident)) {
-    KMXPLUS_PRINTF((" (invalid section name) \n"));
+  if (size > length) {
+    debug_println("size > length");
+    return false;
+  }
+  if (!validate_section_name(ident)) {
+    debug_println("invalid section name");
     return false;
   }
-  KMXPLUS_PRINTF((": (%X) size 0x%X\n", hdr->ident, hdr->size));
   return true;
 }
 
-static bool
-validate_kmxplus_keys(const uint8_t* /*data*/, const COMP_KMXPLUS_KEYS* keys) {
-  if(keys == nullptr) {
-    KMXPLUS_PRINTF(("! could not load 'keys' section\n"));
-    return false;
-  }
-  if(!validate_kmxplus_header((const COMP_KMXPLUS_HEADER*)keys)) {
+bool
+COMP_KMXPLUS_KEYS::valid(KMX_DWORD _kmn_unused(length)) const {
+  KMXPLUS_PRINTF((" count: #0x%X\n", this->count));
+  if (header.size < sizeof(*this)+(sizeof(entries[0])*count)) {
```

Comment:
Should this be a `!=`? Would there be a reason why `header.size` is greater than the sum of these components? Same applies in other `valid` functions :grin:

Reply:
ooo yes.

Reply:
forward compat would be the only reason for ≠

Reply:
actually my hand assembled tiny would also have a ≠ because there's extra padding at the end of the strs.